waku/waku_sync/protocol.nim

```diff
@@ -0,0 +1,311 @@
+  storage: Storage # Negentropy protocol storage 
+  peerManager: PeerManager
+  maxFrameSize: int
+
+  syncInterval: timer.Duration # Time between each syncronisation attempt
+  relayJitter: Duration # Time delay until all messages are mostly received network wide
+  transferCallBack: Option[TransferCallback] # Callback for message transfer.
+
+  pruneCallBack: Option[PruneCallBack] # Callback with the result of the archive query
+  pruneStart: Timestamp # Last pruning start timestamp
+  pruneOffset: timer.Duration # Offset to prune a bit more than necessary. 
+
+  periodicSyncFut: Future[void]
+  periodicPruneFut: Future[void]
+
+proc storageSize*(self: WakuSync): int =
+  self.storage.len
+
+proc ingessMessage*(self: WakuSync, pubsubTopic: PubsubTopic, msg: WakuMessage) =
+  if msg.ephemeral:
+    return
+  #TODO: Do we need to check if this message has already been ingessed? 
+  # because what if messages is received via gossip and sync as well?
+  # Might 2 entries to be inserted into storage which is inefficient.
+  let msgHash: WakuMessageHash = computeMessageHash(pubsubTopic, msg)
+
+  trace "inserting message into storage ",
+    hash = msgHash.toHex(), timestamp = msg.timestamp
+
+  if self.storage.insert(msg.timestamp, msgHash).isErr():
+    debug "failed to insert message ", hash = msgHash.toHex()
+
+proc calculateRange(jitter: Duration, syncRange: Duration = 1.hours): (int64, int64) =
+  var now = getNowInNanosecondTime()
+
+  # Because of message jitter inherent to Relay protocol
+  now -= jitter.nanos
+
+  let range = syncRange.nanos
+
+  let start = now - range
+  let `end` = now
+
+  return (start, `end`)
+
+proc request(
+    self: WakuSync, conn: Connection
+): Future[Result[seq[WakuMessageHash], string]] {.async, gcsafe.} =
+  let (start, `end`) = calculateRange(self.relayJitter)
+
+  let initialized =
+    ?clientInitialize(self.storage, conn, self.maxFrameSize, start, `end`)
+
+  debug "sync session initialized",
+    client = self.peerManager.switch.peerInfo.peerId,
+    server = conn.peerId,
+    frameSize = self.maxFrameSize,
+    timeStart = start,
+    timeEnd = `end`
+
+  var hashes: seq[WakuMessageHash]
+  var reconciled = initialized
+
+  while true:
+    let sent = ?await reconciled.send()
+
+    trace "sync payload sent",
+      client = self.peerManager.switch.peerInfo.peerId,
+      server = conn.peerId,
+      payload = reconciled.payload
+
+    let received = ?await sent.listenBack()
+
+    trace "sync payload received",
+      client = self.peerManager.switch.peerInfo.peerId,
+      server = conn.peerId,
+      payload = received.payload
+
+    reconciled = (?received.clientReconcile(hashes)).valueOr:
+      let completed = error # Result[Reconciled, Completed]
+
+      ?await completed.clientTerminate()
+
+      debug "sync session ended gracefully",
+        client = self.peerManager.switch.peerInfo.peerId, server = conn.peerId
+
+      return ok(hashes)
+
+proc sync*(
```

Comment:
Maybe can merge this with function below?
Almost all code seems duplicated...maybe can make peer as optional arg to the Sync method.